### DISCUSSION TITLE
Renovate attempts to upgrade ruby gems that can't be upgraded due to Ruby version

### DISCUSSION THREAD

**Discussion by DawidJanczak** [renovatebot/renovate#21184] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Wanted end result.

Hello,

Renovate recently opens PRs that have a library upgrade in its title and/or description that cannot be done due to Ruby version constraints. As an example consider this PR in the minimal reproduction repo I prepared: https://github.com/DawidJanczak/renovate-ruby-version/pull/5 It mentions upgrading sidekiq to version 7 where that's not possible due to ruby-version being set to 2.6.2 (sidekiq 7 requires ruby 2.7). The upgrade doesn't go through in the end, as illustrated by the diff, but I believe this is a regression - previously sidekiq wouldn't have attempted to be upgraded. Unfortunately I don't know when this got introduced, but it had to be within the last two weeks since we merged a renovate PR on March 13 that did not mention sidekiq whereas the one created today (March 27) does (this is in a separate repo, not the minimal reproduction one that I just created).

### What you tried so far.

Haven't tried anything new since I believe this is a change in renovate's behaviour, but it would be good to confirm this.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: packageFiles with updates
{
  "baseBranch": "main",
  "config": {
    "bundler": [
      {
        "deps": [
          {
            "currentVersion": "6.5.8",
            "datasource": "rubygems",
            "depName": "sidekiq",
            "fixedVersion": "6.5.8",
            "isSingleVersion": true,
            "lockedVersion": "6.5.8",
            "managerData": {
              "lineNumber": 2
            },
            "packageName": "sidekiq",
            "registryUrl": "https://rubygems.org",
            "versioning": "ruby",
            "warnings": [],
            "updates": [
              {
                "bucket": "major",
                "newVersion": "7.0.7",
                "newMajor": 7,
                "newMinor": 0,
                "updateType": "major",
                "isRange": true,
                "isLockfileUpdate": true,
                "branchName": "renovate/sidekiq-7.x-lockfile"
              }
            ]
          }
        ],
        "lockFiles": [
          "Gemfile.lock"
        ],
        "packageFile": "Gemfile",
        "registryUrls": [
          "https://rubygems.org"
        ]
      }
    ],
    "ruby-version": [
      {
        "deps": [
          {
            "currentValue": "2.6.2",
            "datasource": "ruby-version",
            "depName": "ruby",
            "packageName": "ruby",
            "skipReason": "ignored",
            "updates": []
          }
        ],
        "packageFile": ".ruby-version"
      }
    ]
  }
}
```

</details>

**Comment by rarkins** [comment, marked as answer]
The difference in behavior you have noticed is due to this change in defaults in v35:

config: Renovate will now default to updating locked dependency versions. To revert to previous behavior, configure rangeStrategy=replace

However either way Renovate does not yet do the type of deep compatibility support which you are expecting it does.

For supported datasources, there exists the `constraintsFiltering` option, which defaults to none. However, rubygems is not yet a supported datasource for that. You are welcome to submit a _feature request_ Issue for constraintsFiltering support for rubygems and attach your reproduction repo as an example.

**Reply by DawidJanczak** [reply]
Thanks for the reply @rarkins. I find this a little confusing because `rangeStrategy` docs seem to imply this is for situations where a range is defined in the main package file (i.e. `Gemfile` or `package.json`), but in this case no version is mentioned in `Gemfile`. Also out of curiosity what was the previous default if the new one is `replace`?

**Reply by rarkins** [reply]
The new default is `update-lockfile` when `rangeStrategy=auto`. If you want to revert to previous behavior, you need to add `rangeStrategy=replace` to your config.

**Reply by DawidJanczak** [reply]
Right I think I finally get this, but it's a bit tricky to reason about when there's no range involved in the package file. Is it that in this case the range is implied as something like `0..Infinity` (anything matches) or is it more like `= X.Y.Z` (only the specific version that's in the lockfile matches)?

**Reply by rarkins** [reply]
If there's no range in the package file then it's treated as any/infinity and therefore never needs "replacing".

**Reply by DawidJanczak** [reply]
Yeah OK that makes sense, appreciate the explanation @rarkins
